pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -0,0 +1,292 @@
+		endpointCache: newendpointSliceCache(),
+	}
+}
+
+func (e endpointSliceController) Get() cacheHandler {
+	return e.cache
+}
+
+func (e endpointSliceController) updateEDSSlice(c *Controller, slice *discoveryv1alpha1.EndpointSlice, event model.Event) {
+	svcName := slice.Labels[discoveryv1alpha1.LabelServiceName]
+	hostname := kube.ServiceHostname(svcName, slice.Namespace, c.domainSuffix)
+
+	endpoints := make([]*model.IstioEndpoint, 0)
+	if event != model.EventDelete {
+		for _, e := range slice.Endpoints {
+			if e.Conditions.Ready != nil && !*e.Conditions.Ready {
+				// Ignore not ready endpoints
+				continue
+			}
+			for _, a := range e.Addresses {
+				pod := c.pods.getPodByIP(a)
+				if pod == nil {
+					// This can not happen in usual case
+					if e.TargetRef != nil && e.TargetRef.Kind == "Pod" {
+						log.Warnf("Endpoint without pod %s %s.%s", a, svcName, slice.Namespace)
+
+						if c.metrics != nil {
+							c.metrics.AddMetric(model.EndpointNoPod, string(hostname), nil, a)
+						}
+						// TODO: keep them in a list, and check when pod events happen !
+						continue
+					}
+					// For service without selector, maybe there are no related pods
+				}
+
+				var labels map[string]string
+				locality := GetLocalityFromTopology(e.Topology)
+				sa, uid := "", ""
+				if pod != nil {
+					sa = kube.SecureNamingSAN(pod)
+					uid = fmt.Sprintf("kubernetes://%s.%s", pod.Name, pod.Namespace)
+					labels = configKube.ConvertLabels(pod.ObjectMeta)
+				}
+
+				tlsMode := kube.PodTLSMode(pod)
+
+				// EDS and ServiceEntry use name for service port - ADS will need to
+				// map to numbers.
+				for _, port := range slice.Ports {
+					var portNum uint32
+					if port.Port != nil {
+						portNum = uint32(*port.Port)
+					}
+					var portName string
+					if port.Name != nil {
+						portName = *port.Name
+					}
+					endpoints = append(endpoints, &model.IstioEndpoint{
+						Address:         a,
+						EndpointPort:    portNum,
+						ServicePortName: portName,
+						Labels:          labels,
+						UID:             uid,
+						ServiceAccount:  sa,
+						Network:         c.endpointNetwork(a),
+						Locality:        locality,
+						Attributes:      model.ServiceAttributes{Name: svcName, Namespace: slice.Namespace},
+						TLSMode:         tlsMode,
+					})
+				}
+			}
+		}
+	}
+
+	e.endpointCache.Update(hostname, slice.Name, endpoints)
+
+	log.Infof("Handle EDS endpoint %s in namespace %s", svcName, slice.Namespace)
+
+	if features.EnableHeadlessService.Get() {
+		if obj, _, _ := c.services.informer.GetIndexer().GetByKey(kube.KeyFunc(svcName, slice.Namespace)); obj != nil {
+			svc := obj.(*v1.Service)
+			// if the service is headless service, trigger a full push.
+			if svc.Spec.ClusterIP == v1.ClusterIPNone {
+				c.xdsUpdater.ConfigUpdate(&model.PushRequest{Full: true, NamespacesUpdated: map[string]struct{}{slice.Namespace: {}}})
+				return
+			}
+		}
+	}
```

Comment:
nit: move this in the front, so we donot need to process endpoints convert for headless service

Reply:
or we can move it to `AppendInstanceHandler`

Reply:
I moved to AppendInstanceHandler